### PR TITLE
transfer memcpy_h2d from fluid to phi

### PR DESCRIPTION
**Pull request by kangguangli**
### PR types
Others

### PR changes
OPs

### Describe
1. transfer memcpy_h2d cpu/gpu/xpu kernel to phi
2. add phi kernel compatibility for  DataTranferHelper::RunAndConstructOpFuncNode

### CODE REVIEW

File: paddle/fluid/framework/new_executor/data_transfer.cc [hunk 1]
```diff
@@ -110,33 +112,61 @@ void DataTranferHelper::RunAndConstructOpFuncNode(
...
+  if (phi_kernel_map.size() > 0) {
+    auto phi_kernel_key = op_with_kernel->ChoosePhiKernel(exec_ctx);
+    VLOG(6) << "phi_kernel_key " << phi_kernel_key << "\n";
+
+    PADDLE_ENFORCE_EQ(op_with_kernel->PhiKernel()->IsValid(),
+                      true,
+                      platform::errors::PreconditionNotMet(
+                          "this function was expected to construct"
+                          "data transfer op, but %s was passed.",
+                          op_with_kernel->Type()));
```
Comment (zhiqiu) [inline paddle/fluid/framework/new_executor/data_transfer.cc]: To make it clear. -> "The xx op has no valid phi kernel"

File: paddle/phi/kernels/memcpy_h2d_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+template <typename T, typename Context>
+void MemcpyH2DKernel(const Context& dev_ctx,
+                     const DenseTensor& x,
+                     int dst_place_type,
+                     DenseTensor* out) {
+  PADDLE_ENFORCE_GE(
+      dst_place_type,
+      0,
+      paddle::platform::errors::OutOfRange(
+          "dst_place_type only support 0-3 now: %d", dst_place_type));
```
Comment (zhiqiu) [inline paddle/phi/kernels/memcpy_h2d_kernel.cc]:
```suggestion
          "dst_place_type only supports 0-3, but got: %d", dst_place_type));
```

Comment (kangguangli) [reply paddle/phi/kernels/memcpy_h2d_kernel.cc]: done

File: paddle/fluid/framework/new_executor/data_transfer.cc (hunk 1 above)
Comment (kangguangli) [reply paddle/fluid/framework/new_executor/data_transfer.cc]: done

File: paddle/phi/kernels/memcpy_h2d_kernel.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+template <typename T, typename Context>
+void MemcpyH2DKernel(const Context& dev_ctx,
+                     const DenseTensor& x,
+                     int dst_place_type,
+                     DenseTensor* out) {
+  PADDLE_ENFORCE_GE(
+      dst_place_type,
+      0,
+      paddle::platform::errors::OutOfRange(
```
Comment (zhiqiu) [inline paddle/phi/kernels/memcpy_h2d_kernel.cc]: -> phi::errors::xx

File: paddle/phi/kernels/memcpy_h2d_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/memcpy_h2d_kernel.h"
+
+#include "paddle/fluid/framework/tensor_util.h"
```
Comment (zhiqiu) [inline paddle/phi/kernels/memcpy_h2d_kernel.cc]: it can be removed

File: paddle/phi/kernels/memcpy_h2d_kernel.cc (hunk 3 above)
Comment (kangguangli) [reply paddle/phi/kernels/memcpy_h2d_kernel.cc]: done

File: paddle/phi/kernels/memcpy_h2d_kernel.cc (hunk 4 above)
Comment (kangguangli) [reply paddle/phi/kernels/memcpy_h2d_kernel.cc]: done

**Review by zhiqiu — APPROVED** [review]
LGTM, good work!

File: paddle/phi/kernels/memcpy_h2d_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chenwhql) [inline paddle/phi/kernels/memcpy_h2d_kernel.cc:1]: 后面看要不要这个文件名改改，比如去掉h2d，把d2h的kernel也放到这个文件中，减少几个文件

Comment (kangguangli) [reply paddle/phi/kernels/memcpy_h2d_kernel.cc:1]: ok

**Pull request merged**